enterprise/server/remote_execution/blockio/blockio.go

```diff
@@ -220,12 +235,50 @@ func NewCOWStore(chunks []*Chunk, chunkSizeBytes, totalSizeBytes int64, dataDir
 		dirty:          make(map[int64]bool, 0),
 		dataDir:        dataDir,
 		copyBuf:        make([]byte, chunkSizeBytes),
+		zeroBuf:        make([]byte, chunkSizeBytes),
 		chunkSizeBytes: chunkSizeBytes,
 		totalSizeBytes: totalSizeBytes,
 		ioBlockSize:    int64(stat.Sys().(*syscall.Stat_t).Blksize),
 	}, nil
 }
 
+// GetPageAddress returns a memory address for the given byte offset which can
+// be used to handle a page fault with userfaultfd.
+//
+// If reading a lazily mmapped chunk, this will cause the chunk to be mmapped
+// so that the returned address is valid.
+//
+// If write is set to true, and the page is not dirty, then a copy is first
+// performed so that the returned chunk can be written to without modifying
+// readonly chunks.
+func (c *COWStore) GetPageAddress(address uintptr, write bool) (uintptr, error) {
+	chunkStartOffset := c.chunkStartOffset(int64(address))
+	chunkRelativeAddress := address - uintptr(chunkStartOffset)
+	if write && !c.dirty[chunkStartOffset] {
+		if err := c.copyChunk(chunkStartOffset); err != nil {
+			return 0, status.WrapError(err, "copy chunk")
+		}
+	}
+	chunk := c.chunks[chunkStartOffset]
+	if chunk == nil {
+		// No data (yet); map into our static zero-filled buf. Note that this
+		// can only happen for reads, since for writes we call copyChunk above.
+		ptr := uintptr(unsafe.Pointer(&c.zeroBuf[0]))
```

Comment:
nit: What do you think of adding a helper function for uintptr(unsafe.Pointer() so it's more clear that is just getting the start address of an array?